mono_repo/test/yaml_write_test.dart

```diff
@@ -0,0 +1,695 @@
+import 'package:yaml/yaml.dart' as y;
+
+import 'shared.dart';
+
+void _asciiTest(List<int> bytes) {
+  test(bytes.toString(), () {
+    var source = ascii.decode(bytes);
+    var output = toYaml(source);
+    var yaml = y.loadYaml(output);
+    expect(yaml, source, reason: bytes.toString());
+  });
+}
+
+void main() {
+  group('ascii', () {
+    for (var i = 0; i < 128; i++) {
+      _asciiTest([i]);
+    }
+  });
+
+  group('naughty strings', () {
+    for (var source in _stringEscapeSamples.followedBy(_naughtyStrings)) {
+      _testRoundTrip(source);
+    }
+  });
+
+  for (var item in [
+    '',
+    'string',
+    '"double quotes"',
+    "'single quotes'",
+    '\"double\" and \'single\' quotes',
+    '\n',
+    '\t\v\r',
+    '---', // special marker in Yaml
+    '...', // special marker in Yaml
+
+    // booleans
+    true,
+    false,
+    'TRUE',
+    'FALSE',
+    'True',
+    'False',
+    'Yes',
+    'No',
+
+    // null
+    null,
+    'Null',
+    'NULL',
+
+    // numbers
+    0,
+    0.0,
+    1,
+    -1,
+    3.14,
+    -3.14,
+    1.0,
+    double.maxFinite,
+    double.minPositive,
+    '+1',
+    '.1',
+    '+.1',
+
+    // Maps
+    {},
+    {
+      'null': null,
+      'bool': true,
+      'int': 1,
+      'double': 3.14,
+      'string': 'string',
+      'list': [],
+      'map': {}
+    },
+    {null: 'null', true: 'bool', 1: 'int', 3.14: 'double', 'string': 'string'},
+    {
+      'stages': [1, 2, 3]
+    },
+    {'list': []},
+    {'map': {}},
+    {
+      'jobs': {
+        'include': [
+          {'item1': 'value1', 'item2': 'value2'},
+          {'item1': 'value1', 'item2': 'value2'},
+        ]
+      }
+    },
+
+    // Lists
+    [],
+    [[]],
+    [null, true, false, 1, 3.14, 'string', [], {}],
+    [
+      {'test': 1},
+      {'test': 2}
+    ]
+  ]) {
+    _testRoundTrip(item);
+  }
+
+  group('unsupported', () {
+    group('non-simple map keys', () {
+      for (var key in [[], {}]) {
+        test('`${jsonEncode(key)}`', () {
+          expect(() => toYaml({key: 'value'}), throwsArgumentError);
+        });
+      }
+    });
+  });
+
+  group('config files', () {
+    for (var entry in {'1': testConfig1, '2': testConfig2}.entries) {
+      test(entry.key, () {
+        var decoded = y.loadYaml(entry.value);
+        var output = toYaml(decoded);
+        printOnFailure(['# start yaml', output, '# end yaml'].join('\n'));
+        var roundTrip = y.loadYaml(output);
+        expect(roundTrip, decoded);
+      });
+    }
+  });
+}
+
+void _testRoundTrip(Object source) {
+  String testTitle;
+  try {
+    testTitle = jsonEncode(source);
+  } on JsonUnsupportedObjectError {
+    testTitle = source.toString();
+  }
+
+  test(testTitle, () {
+    var output = toYaml(source);
+    printOnFailure(['# start yaml', output, '# end yaml'].join('\n'));
+    var yaml = y.loadYaml(output);
+    expect(yaml, source);
+  });
+
+  if (source is! String) {
+    _testRoundTrip(source.toString());
```

Comment:
what is the purpose of this recursive call? can you make it more explicit?

Reply:
ack